presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayPositionFunction.java

```diff
@@ -13,93 +13,38 @@
-    private static final MethodHandle METHOD_HANDLE_SLICE = methodHandle(ArrayPositionFunction.class, "arrayPosition", Type.class, MethodHandle.class, Block.class, Slice.class);
-    private static final MethodHandle METHOD_HANDLE_OBJECT = methodHandle(ArrayPositionFunction.class, "arrayPosition", Type.class, MethodHandle.class, Block.class, Object.class);
-
-    public ArrayPositionFunction()
-    {
-        super("array_position", ImmutableList.of(comparableTypeParameter("E")), ImmutableList.of(), "bigint", ImmutableList.of("array(E)", "E"));
-    }
-
-    @Override
-    public boolean isHidden()
-    {
-        return false;
-    }
-
-    @Override
-    public boolean isDeterministic()
-    {
-        return true;
-    }
-
-    @Override
-    public String getDescription()
-    {
-        return "Returns the position of the first occurrence of the given value in array (or 0 if not found)";
-    }
-
-    @Override
-    public ScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
-    {
-        Type type = boundVariables.getTypeVariable("E");
-        MethodHandle equalMethodHandle = functionRegistry.getScalarFunctionImplementation(internalOperator(EQUAL, BOOLEAN, ImmutableList.of(type, type))).getMethodHandle();
-        MethodHandle arrayPositionMethodHandle;
-        if (type.getJavaType() == boolean.class) {
-            arrayPositionMethodHandle = METHOD_HANDLE_BOOLEAN;
-        }
-        else if (type.getJavaType() == long.class) {
-            arrayPositionMethodHandle = METHOD_HANDLE_LONG;
-        }
-        else if (type.getJavaType() == double.class) {
-            arrayPositionMethodHandle = METHOD_HANDLE_DOUBLE;
-        }
-        else if (type.getJavaType() == Slice.class) {
-            arrayPositionMethodHandle = METHOD_HANDLE_SLICE;
-        }
-        else {
-            arrayPositionMethodHandle = METHOD_HANDLE_OBJECT;
-        }
-        return new ScalarFunctionImplementation(false, ImmutableList.of(false, false), arrayPositionMethodHandle.bindTo(type).bindTo(equalMethodHandle), isDeterministic());
-    }
+    private ArrayPositionFunction() {}
 
-    @UsedByGeneratedCode
-    public static long arrayPosition(Type type, MethodHandle equalMethodHandle, Block array, boolean element)
+    @TypeParameter("E")
+    @SqlType(StandardTypes.BIGINT)
+    public static long arrayPosition(@TypeParameter("E") Type type, @OperatorDependency(operator = EQUAL, returnType = StandardTypes.BOOLEAN, argumentTypes = {"E", "E"}) MethodHandle equalMethodHandle, @SqlType("array(E)") Block array, @SqlType("E") Object element)
     {
-        int size = array.getPositionCount();
-        for (int i = 0; i < size; i++) {
+        for (int i = 0; i < array.getPositionCount(); ++i) {
             if (!array.isNull(i)) {
-                boolean arrayValue = type.getBoolean(array, i);
+                Object arrayValue = readNativeValue(type, array, i);
```

Comment:
This is going to hurt performance. `readNativeValue` has a branch in it, so change it back to the way it was.